src/contracts/core/AllocationManager.sol

```diff
@@ -0,0 +1,731 @@
+     * @param operator address to modify allocations for
+     * @param allocations array of magnitude adjustments for multiple strategies and corresponding operator sets
+     * @param operatorSignature signature of the operator if msg.sender is not the operator
+     * @dev Updates freeMagnitude for the updated strategies
+     * @dev Must be called by the operator or with a valid operator signature
+     * @dev For each allocation, allocation.operatorSets MUST be ordered in ascending order according to the
+     * encoding of the operatorSet. This is to prevent duplicate operatorSets being passed in. The easiest way to ensure
+     * ordering is to sort allocated operatorSets by address first, and then sort for each avs by ascending operatorSetIds.
+     */
+    function modifyAllocations(
+        address operator,
+        MagnitudeAllocation[] calldata allocations,
+        SignatureWithSaltAndExpiry calldata operatorSignature
+    ) external {
+        if (msg.sender != operator) {
+            _verifyOperatorSignature(operator, allocations, operatorSignature);
+        }
+        require(delegation.isOperator(operator), OperatorNotRegistered());
+
+        (bool isSet, uint32 delay) = allocationDelay(operator);
+        require(isSet, UninitializedAllocationDelay());
+
+        // effect timestamp for allocations to take effect. This is configurable by operators
+        uint32 allocationEffectTimestamp = uint32(block.timestamp) + delay;
+        // completable timestamp for deallocations
+        uint32 deallocationCompletableTimestamp = uint32(block.timestamp) + DEALLOCATION_DELAY;
+
+        for (uint256 i = 0; i < allocations.length; ++i) {
+            // 1. For the given (operator,strategy) clear all pending free magnitude for the strategy and update freeMagnitude
+            _updateFreeMagnitude({
+                operator: operator,
+                strategy: allocations[i].strategy,
+                numToComplete: type(uint16).max
+            });
+
+            // 2. Check current totalMagnitude matches expected value. This is to check for slashing race conditions
+            // where an operator gets slashed from an operatorSet and as a result all the configured allocations have larger
+            // proprtional magnitudes relative to each other.
+            uint64 currentTotalMagnitude = _getLatestTotalMagnitude(operator, allocations[i].strategy);
+            require(currentTotalMagnitude == allocations[i].expectedTotalMagnitude, InvalidExpectedTotalMagnitude());
+
+            // 3. set allocations for the strategy after updating freeMagnitude
+            _modifyAllocations({
+                operator: operator,
+                allocation: allocations[i],
+                allocationEffectTimestamp: allocationEffectTimestamp,
+                deallocationCompletableTimestamp: deallocationCompletableTimestamp
+            });
```

Comment:
Hidden nested for loop. Might want to pull the for loop in this method to the surface